astropy/coordinates/matching.py

```diff
@@ -74,6 +74,9 @@ def match_coordinates_3d(matchcoord, catalogcoord, nthneighbor=1, storekdtree='k
     matchxyz = matchcoord.cartesian.xyz.to(catunit)
 
     matchflatxyz = matchxyz.reshape((3, np.prod(matchxyz.shape) // 3))
+    # Querying NaN returns garbage
+    if np.isnan(matchflatxyz.value).sum() > 0:
```

Comment:
Here and below, I would use `.any()` instead of `.sum()`: for very large arrays, `.any()` will be significantly faster because it doesn't have yo go through the entire array.
```suggestion
    if np.isnan(matchflatxyz.value).any():
```